File: components/pentode.py
# components/pentode.py
import numpy as np
from components.component import Component

class Pentode(Component):
    def __init__(self, name: str, plate_node: str, grid_node: str, screen_grid_node: str, suppressor_grid_node: str, cathode_node: str,
                 mu: float = 100.0, Kp: float = 1.0, Ex: float = 1.0, Kg1: float = 1.0, Kg2: float = 1.0, Vg_offset: float = 0.0):
        """
        Inizializza un Pentodo usando un modello tipo Koren esteso (o simile).
        Args:
            name (str): Nome univoco dell'istanza.
            plate_node (str): Nodo della placca (anodo).
            grid_node (str): Nodo della griglia di controllo (g1).
            screen_grid_node (str): Nodo della griglia schermo (g2).
            suppressor_grid_node (str): Nodo della griglia soppressore (g3).
            cathode_node (str): Nodo del catodo.
            mu (float): Fattore di amplificazione (per g1).
            Kp (float): Parametro di permeabilità.
            Ex (float): Esponente.
            Kg1 (float): Parametro di linearità della griglia di controllo.
            Kg2 (float): Parametro di linearità della griglia schermo.
            Vg_offset (float): Offset di tensione per la griglia di controllo.
        """
        super().__init__(name, plate_node, grid_node, screen_grid_node, suppressor_grid_node, cathode_node)
        self.pin_names = ('plate', 'grid', 'screen_grid', 'suppressor_grid', 'cathode') # Per mappatura nodi con nome
        self.mu = mu
        self.Kp = Kp
        self.Ex = Ex
        self.Kg1 = Kg1
        self.Kg2 = Kg2
        self.Vg_offset = Vg_offset

    def calculate_plate_current(self, Vgk: float, Vpk: float, Vg2k: float, Vg3k: float) -> float:
        """
        Calcola la corrente di placca (Ip) per un pentodo.
        Questo è un modello molto semplificato; i modelli pentodo reali sono complessi.
        Args:
            Vgk (float): Tensione Griglia-Catodo.
            Vpk (float): Tensione Placca-Catodo.
            Vg2k (float): Tensione Griglia Schermo-Catodo.
            Vg3k (float): Tensione Griglia Soppressore-Catodo.
        Returns:
            float: Corrente di placca in Ampere.
        """
        # Per semplicità, questo modello ignora Vg3k e usa un approccio simile al triodo,
        # ma con l'influenza della griglia schermo.
        # Un modello pentodo più accurato richiederebbe equazioni più complesse
        # che modellano la partizione di corrente tra placca e griglia schermo.

        if Vpk <= 0:
            return 0.0 # Nessuna corrente se la placca non è polarizzata positivamente

        # Calcolo della tensione di griglia equivalente che controlla la corrente totale
        # Questo è un punto di semplificazione, i modelli reali sono più complessi
        effective_grid_voltage = (Vgk + self.Vg_offset) + (Vg2k / self.mu)

        if effective_grid_voltage <= 0:
            return 0.0 # Cutoff

        # La corrente di placca è influenzata dalla tensione di placca e di griglia schermo
        # Questo è un modello comportamentale, non un modello fisico completo di pentodo.
        # Per un modello più rigoroso, si dovrebbe considerare la partizione di corrente
        # tra placca e griglia schermo e la saturazione di placca.
        
        # Un approccio comune è usare una funzione di saturazione per la placca (es. arctan, tanh)
        # per modellare il "ginocchio" della curva di placca.
        # Per ora, usiamo una dipendenza lineare dalla tensione di placca per semplicità,
        # ma questo è un punto da migliorare per fedeltà.
        
        # La corrente totale (placca + griglia schermo) dipende principalmente da Vgk e Vg2k
        total_current = self.Kp * (effective_grid_voltage)**self.Ex

        # Partizione della corrente (molto semplificata: la corrente di placca è una frazione della totale)
        # In un pentodo, la corrente di placca è quasi indipendente da Vpk una volta in saturazione.
        # Qui, usiamo un fattore che dipende da Vpk per simulare una leggera dipendenza.
        # Questo è un placeholder e andrebbe sostituito con un modello pentodo più specifico (es. Koren Pentode, o modelli basati su curve).
        plate_current_factor = np.tanh(Vpk / 50.0) # Esempio di funzione di saturazione per Vpk

        Ip = total_current * plate_current_factor
        
        # Aggiungi un limite per evitare valori irrealistici
        if Ip > 1.0:
            Ip = 1.0

        return Ip

    def get_stamps(self, num_total_equations: int, dt: float, current_solution_guess: np.ndarray, prev_solution: np.ndarray, time: float):
        """
        Per i componenti non lineari, get_stamps restituisce matrici/vettori vuoti.
        Il loro contributo è gestito direttamente nel _system_equations del solutore.
        """
        return np.zeros((num_total_equations, num_total_equations)), np.zeros(num_total_equations)

File: components/Pentode.cpp
// components/Pentode.h
#ifndef PENTODE_H
#define PENTODE_H

#include "Component.h" // Include la classe base Component
#include <Eigen/Dense>   // Per le operazioni con matrici e vettori Eigen
#include <string>        // Per std::string
#include <vector>        // Per std::vector
#include <map>           // Per std::map (per getNodeIndex)

/**
 * @class Pentode
 * @brief Rappresenta un componente pentodo per simulazioni di circuiti MNA.
 *
 * Questa classe modella un pentodo con i suoi cinque terminali principali:
 * Griglia di Controllo (Control Grid), Placca (Plate), Catodo (Cathode),
 * Griglia Schermo (Screen Grid), e Griglia Soppressore (Suppressor Grid).
 * Implementa i metodi per applicare gli "stamps" alla matrice MNA e al vettore RHS,
 * utilizzando un modello non lineare semplificato adatto all'analisi transitoria
 * con iterazione di Newton-Raphson.
 */
class Pentode : public Component {
public:
    // Nomi dei nodi per i terminali del pentodo
    std::string node_control_grid;
    std::string node_plate;
    std::string node_cathode;
    std::string node_screen_grid;
    std::string node_suppressor_grid;

    // Parametri del modello semplificato del pentodo
    double K_p_;        ///< Parametro di transconduttanza (simile a Kp in MOSFET)
    double V_cutoff_;   ///< Tensione di cutoff della griglia di controllo
    double lambda_;     ///< Parametro di modulazione della lunghezza del canale (per dipendenza dalla placca)
    double mu_sg_;      ///< Fattore di amplificazione della griglia schermo (quanto Vsg influenza Iplate rispetto a Vcg)
    double I_s_ratio_;  ///< Rapporto tra corrente di griglia schermo e corrente di placca (semplificato)

    // Variabili di stato per l'iterazione di Newton-Raphson (tensioni del passo precedente)
    // Queste sono le tensioni dei nodi calcolate all'iterazione precedente del Newton-Raphson.
    double prev_Vcg_;   ///< Tensione precedente della griglia di controllo
    double prev_Vp_;    ///< Tensione precedente della placca
    double prev_Vsg_;   ///< Tensione precedente della griglia schermo
    double prev_Vsup_;  ///< Tensione precedente della griglia soppressore

    /**
     * @brief Costruttore per il componente Pentode.
     * @param name Il nome univoco del pentodo.
     * @param control_grid_node Il nome del nodo della griglia di controllo.
     * @param plate_node Il nome del nodo della placca.
     * @param cathode_node Il nome del nodo del catodo.
     * @param screen_grid_node Il nome del nodo della griglia schermo.
     * @param suppressor_grid_node Il nome del nodo della griglia soppressore.
     * @param K_p Parametro di transconduttanza.
     * @param V_cutoff Tensione di cutoff.
     * @param lambda Parametro lambda per la modulazione della lunghezza del canale.
     * @param mu_sg Fattore di amplificazione della griglia schermo.
     * @param I_s_ratio Rapporto corrente griglia schermo/placca.
     * @param tolerance_percent Percentuale di tolleranza opzionale.
     */
    Pentode(const std::string& name,
            const std::string& control_grid_node, const std::string& plate_node,
            const std::string& cathode_node, const std::string& screen_grid_node,
            const std::string& suppressor_grid_node,
            double K_p, double V_cutoff, double lambda, double mu_sg, double I_s_ratio,
            double tolerance_percent = 0.0);

    /**
     * @brief Crea una copia profonda dell'oggetto Pentode.
     * @return Un puntatore a un nuovo oggetto Pentode, che è una copia dell'istanza corrente.
     */
    Component* clone() const override { return new Pentode(*this); }

    /**
     * @brief Applica gli "stamps" del pentodo alla matrice MNA (A) e al vettore (b).
     *
     * Questo metodo implementa un modello non lineare semplificato per il pentodo,
     * calcolando le correnti di placca e griglia schermo e le loro conduttanze dinamiche
     * per l'iterazione di Newton-Raphson.
     *
     * @param A La matrice MNA a cui vengono applicati gli stamps.
     * @param b Il vettore lato destro MNA a cui vengono applicati gli stamps.
     * @param x_current_guess La stima corrente per le tensioni dei nodi e le correnti di ramo.
     * @param prev_solution La soluzione dal passo temporale precedente (non usata direttamente per il modello statico non lineare).
     * @param time Il tempo di simulazione corrente.
     * @param dt La dimensione del passo temporale.
     */
    void getStamps(
        Eigen::MatrixXd& A, Eigen::VectorXd& b,
        const Eigen::VectorXd& x_current_guess, const Eigen::VectorXd& prev_solution,
        double time, double dt
    ) override;

    /**
     * @brief Aggiorna lo stato interno del pentodo.
     *
     * Questo metodo memorizza le tensioni dei nodi attuali come "precedenti"
     * per la prossima iterazione di Newton-Raphson nel metodo getStamps.
     *
     * @param v_curr La tensione corrente attraverso il componente (non usata direttamente).
     * @param i_curr La corrente corrente che attraversa il componente (non usata direttamente).
     */
    void updateState(double v_curr, double i_curr) override;
};

#endif // PENTODE_H
